velox/connectors/hive/storage_adapters/hdfs/tests/HdfsFileSystemTest.cpp

```diff
@@ -0,0 +1,331 @@
+  }
+}
+
+TEST_F(HdfsFileSystemTest, multipleThreadsWithReadFile) {
+  startThreads = false;
+  struct hdfsBuilder* builder = hdfsNewBuilder();
+  hdfsBuilderSetNameNode(builder, "localhost");
+  hdfsBuilderSetNameNodePort(builder, 7878);
+  auto hdfs = hdfsBuilderConnect(builder);
+  HdfsReadFile readFile(hdfs, destinationPath);
+  std::vector<std::thread> threads;
+  std::mt19937 generator(std::random_device{}());
+  std::vector<int> sleepTimesInMicroseconds = {0, 500, 50000};
+  std::uniform_int_distribution<std::size_t> distribution(
+      0, sleepTimesInMicroseconds.size() - 1);
+  for (int i = 0; i < 25; i++) {
+    auto thread = std::thread(
+        [&readFile, &distribution, &generator, &sleepTimesInMicroseconds] {
+          int index = distribution(generator);
+          while (!HdfsFileSystemTest::startThreads) {
+            std::this_thread::yield();
+          }
+          std::this_thread::sleep_for(
+              std::chrono::microseconds(sleepTimesInMicroseconds[index]));
+          readData(&readFile);
+        });
+    threads.emplace_back(std::move(thread));
+  }
+  startThreads = true;
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+TEST_F(HdfsFileSystemTest, multipleThreadsWithFileSystem) {
+  startThreads = false;
+  facebook::velox::filesystems::registerHdfsFileSystem();
+  std::unordered_map<std::string, std::string> configurationValues(
+      {{"hdfs_host", "localhost"}, {"hdfs_port", "7878"}});
+  auto memConfig =
+      std::make_shared<const core::MemConfig>(std::move(configurationValues));
+  auto hdfsFileSystem =
+      filesystems::getFileSystem(fullDestinationPath, memConfig);
+
+  std::vector<std::thread> threads;
+  std::mt19937 generator(std::random_device{}());
+  std::vector<int> sleepTimesInMicroseconds = {0, 500, 50000};
+  std::uniform_int_distribution<std::size_t> distribution(
+      0, sleepTimesInMicroseconds.size() - 1);
+  for (int i = 0; i < 25; i++) {
+    auto thread = std::thread([&hdfsFileSystem,
+                               &distribution,
+                               &generator,
+                               &sleepTimesInMicroseconds] {
+      int index = distribution(generator);
+      while (!HdfsFileSystemTest::startThreads) {
+        std::this_thread::yield();
+      }
+      std::this_thread::sleep_for(
+          std::chrono::microseconds(sleepTimesInMicroseconds[index]));
+      auto readFile = hdfsFileSystem->openFileForRead(fullDestinationPath);
+      readData(readFile.get());
+    });
+    threads.emplace_back(std::move(thread));
+  }
+  startThreads = true;
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+TEST_F(HdfsFileSystemTest, readFailures) {
+  struct hdfsBuilder* builder = hdfsNewBuilder();
+  hdfsBuilderSetNameNode(builder, "localhost");
+  hdfsBuilderSetNameNodePort(builder, 7878);
+  auto hdfs = hdfsBuilderConnect(builder);
+  HdfsReadFile readFile(hdfs, destinationPath);
+  verifyFailures(&readFile);
```

Comment:
should we inline `verifyFailures` since it's being used only once?

Reply:
I think abstracting into a method is ok, there is a lot going on there. And putting it all in the test method would make it harder to understand whats going on. Plus there are still some methods that are not supported by the hdfs adapter so I think laying out a structure for the future test case writer is beneficial.

Reply:
sounds good.